File: gsd09-web-crawler/html-downloader/src/main/java/htmldownloader/utils/LinkExtractionUtils.java
package htmldownloader.utils;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Slf4j
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class LinkExtractionUtils {

    private static final Pattern HREF_PATTERN = Pattern.compile(
            "<a\\s+[^>]*href\\s*=\\s*[\"']([^\"']+)[\"'][^>]*>",
            Pattern.CASE_INSENSITIVE | Pattern.DOTALL
    );

    public static List<String> extractLinks(String html, String baseUrl) {
        if (html == null || html.trim().isEmpty()) {
            return new ArrayList<>();
        }

        Set<String> uniqueUrls = new HashSet<>();
        Matcher matcher = HREF_PATTERN.matcher(html);

        while (matcher.find()) {
            String href = matcher.group(1);
            String absoluteUrl = convertToAbsoluteUrl(href, baseUrl);

            if (absoluteUrl != null && isValidUrl(absoluteUrl)) {
                uniqueUrls.add(absoluteUrl);
            }
        }

        log.info("[LinkExtractionUtils.extractLinks] Extracted {} unique links from HTML", uniqueUrls.size());
        return new ArrayList<>(uniqueUrls);
    }

    private static String convertToAbsoluteUrl(String href, String baseUrl) {
        try {
            if (href.startsWith("http://") || href.startsWith("https://")) {
                return href;
            }

            URL base = new URL(baseUrl);

            if (href.startsWith("/")) {
                return base.getProtocol() + "://" + base.getHost() +
                       (base.getPort() != -1 ? ":" + base.getPort() : "") + href;
            }

            if (href.startsWith("../")) {
                return new URL(base, href).toString();
            }

            if (href.startsWith("./")) {
                return new URL(base, href).toString();
            }

            if (href.startsWith("#") || href.startsWith("javascript:") || href.startsWith("mailto:")) {
                return null;
            }

            return new URL(base, href).toString();

        } catch (MalformedURLException e) {
            log.warn("[LinkExtractionUtils.convertToAbsoluteUrl] Failed to convert href to absolute URL: href={}, baseUrl={}", href, baseUrl);
            return null;
        }
    }

    private static boolean isValidUrl(String url) {
        try {
            URL urlObj = new URL(url);
            return "http".equals(urlObj.getProtocol()) || "https".equals(urlObj.getProtocol());
        } catch (MalformedURLException e) {
            return false;
        }
    }

    public static String extractRootDomain(String url) {
        try {
            URL urlObj = new URL(url);
            String host = urlObj.getHost();

            if (host.startsWith("www.")) {
                return host.substring(4);
            }

            return host;
        } catch (MalformedURLException e) {
            log.warn("[LinkExtractionUtils.extractRootDomain] Failed to extract root domain from URL: {}", url);
            return null;
        }
    }
}

File: sdip09-web-crawler-ktor/html-downloader/src/main/kotlin/htmldownloader/util/LinkExtractionUtils.kt
package htmldownloader.util

import org.slf4j.LoggerFactory
import java.net.MalformedURLException
import java.net.URL

object LinkExtractionUtils {
    private val log = LoggerFactory.getLogger(LinkExtractionUtils::class.java)

    private val HREF_PATTERN = Regex(
        """<a\s+[^>]*href\s*=\s*["']([^"']+)["'][^>]*>""",
        RegexOption.IGNORE_CASE
    )

    fun extractLinks(html: String?, baseUrl: String): List<String> {
        if (html.isNullOrBlank()) {
            return emptyList()
        }

        val uniqueUrls = mutableSetOf<String>()

        HREF_PATTERN.findAll(html).forEach { matchResult ->
            val href = matchResult.groupValues[1]
            val absoluteUrl = convertToAbsoluteUrl(href, baseUrl)

            if (absoluteUrl != null && isValidUrl(absoluteUrl)) {
                uniqueUrls.add(absoluteUrl)
            }
        }

        log.info("[LinkExtractionUtils.extractLinks] Extracted {} unique links from HTML", uniqueUrls.size)
        return uniqueUrls.toList()
    }

    private fun convertToAbsoluteUrl(href: String, baseUrl: String): String? {
        return try {
            when {
                href.startsWith("http://") || href.startsWith("https://") -> href
                href.startsWith("#") || href.startsWith("javascript:") || href.startsWith("mailto:") -> null
                else -> {
                    val base = URL(baseUrl)
                    when {
                        href.startsWith("/") -> {
                            val port = if (base.port != -1) ":${base.port}" else ""
                            "${base.protocol}://${base.host}$port$href"
                        }
                        else -> URL(base, href).toString()
                    }
                }
            }
        } catch (e: MalformedURLException) {
            log.warn("[LinkExtractionUtils.convertToAbsoluteUrl] Failed to convert href: href={}, baseUrl={}", href, baseUrl)
            null
        }
    }

    private fun isValidUrl(url: String): Boolean {
        return try {
            val urlObj = URL(url)
            urlObj.protocol == "http" || urlObj.protocol == "https"
        } catch (e: MalformedURLException) {
            false
        }
    }

    fun extractRootDomain(url: String): String? {
        return try {
            val urlObj = URL(url)
            val host = urlObj.host
            if (host.startsWith("www.")) {
                host.substring(4)
            } else {
                host
            }
        } catch (e: MalformedURLException) {
            log.warn("[LinkExtractionUtils.extractRootDomain] Failed to extract root domain from URL: {}", url)
            null
        }
    }
}
